sno/structure.py

```diff
@@ -18,6 +18,126 @@
 L = logging.getLogger("sno.structure")
 
 
+def fast_import_tables(
+    repo, sources, max_pack_size="2G", limit=None, message=None, *, version
+):
+    for path, source in sources.items():
+        if not source.table:
+            raise ValueError("No table specified")
+
+        if not repo.is_empty:
+            if path in repo.head.peel(pygit2.Tree):
+                raise ValueError(f"{path}/ already exists")
+
+    click.echo("Starting git-fast-import...")
+    p = subprocess.Popen(
+        [
+            "git",
+            "fast-import",
+            "--date-format=now",
+            "--done",
+            "--stats",
+            f"--max-pack-size={max_pack_size}",
+        ],
+        cwd=repo.path,
+        stdin=subprocess.PIPE,
+    )
+    try:
+        user = repo.default_signature
+
+        if message is None:
+            if len(sources) == 1:
+                for path, source in sources.items():
+                    message = f"Import from {Path(source.source).name} to {path}/"
+            else:
+                message = f"Import {len(sources)} datasets from '{Path(source.source).name}':\n"
+                for path, source in sources.items():
+                    if path == source.table:
+                        message += f"\n* {path}/"
+                    else:
+                        message += f"\n* {path} (from {source.table})"
+
+        header = (
+            "commit refs/heads/master\n"
+            f"committer {user.name} <{user.email}> now\n"
+            f"data {len(message.encode('utf8'))}\n{message}\n"
+        )
+        p.stdin.write(header.encode("utf8"))
+
+        if not repo.is_empty:
+            # start with the existing tree/contents
+            p.stdin.write(b"from refs/heads/master^0\n")
+        for path, source in sources.items():
+            dataset = DatasetStructure.for_version(version)(tree=None, path=path)
+
+            with source:
+                if limit:
+                    num_rows = min(limit, source.row_count)
+                    click.echo(
+                        f"Importing {num_rows:,d} of {source.row_count:,d} features from {source} to {path}/ ..."
+                    )
+                else:
+                    num_rows = source.row_count
+                    click.echo(
+                        f"Importing {num_rows:,d} features from {source} to {path}/ ..."
+                    )
+
+                t0 = time.monotonic()
+                src_iterator = source.iter_features()
```

Comment:
weird order of statements: src_iterator initialised here but not used until after meta import step is finished.